design-proposals/export-import/volume-export.md

```diff
@@ -0,0 +1,161 @@
+
+## Goals
+* Define VirtualMachineExport API
+* Introduce virtctl export command
+
+## Non Goals
+* Implementation details of internal components (servers/proxies/controllers/etc)
+
+## User Stories
+* As a KubeVirt user, I would like to export a Virtual Machine volume for local debugging/analysis.
+* As a KubeVirt user, I would like to export a Virtual Machine volume as part of migrating a Virtual Machine to a different cluster.
+* As a KubeVirt user, I would like to export a Virtual Machine volume to serve as an offsite backup.
+
+## Repos
+* [KubeVirt](https://github.com/kubevirt/kubevirt)
+
+# Design
+
+## VirtualMachineExport API
+
+The VirtualMachineExport API allows users to download a Virtual Machine volumes from URLs.
+
+### User Flow
+
+1.  Create a VirtualMachineExport resource
+2.  Wait for the `Ready` condition of the VirtualMachineExport to be `True`
+3.  Download a Virtual Machine volume in the desired format listed under `status.links.volumes`
+
+### Export Formats
+
+| Volume Content Type | Export Format |
+| ------------------- | ------------- |
+| kubevirt            | raw           |
+| kubevirt            | tar.gz        |
+| archive             | dir           |
+| archive             | tar.gz        |
+
+#### raw
+
+The http endpoint supports random access to a raw disk image.
+
+#### tar.gz
+
+The http endpoint is a link to a `tar.gz` streaming download.
+
+#### dir
+
+The http endpoint is a golang `http.FileServer` serving up the root of the filesystem.
+
+### Export a VirtualMachine
+
+``` yaml
+apiVersion: export.kubevirt.io/v1alpha1
+kind: VirtualMachineExport
+metadata:
+    name: export-vm1
+spec:
+    source:
+        apiGroup: kubevirt.io
+        kind: VirtualMachine
+        name: vm1
+```
+
+### Export a VirtualMachineSnapshot
+
+``` yaml
+apiVersion: export.kubevirt.io/v1alpha1
+kind: VirtualMachineExport
+metadata:
+    name: export-snap1
+spec:
+    source:
+        apiGroup: snapshot.kubevirt.io
+        kind: VirtualMachineSnapshot
+        name: snap1
+```
+
+### VirtualMachineExport Status
+
+``` yaml
+apiVersion: export.kubevirt.io/v1alpha1
+kind: VirtualMachineExport
+metadata:
+    namespace: ns1
+    name: export-vm1
+spec:
+    ...
+status:
+    phase: Ready # Pending|Ready|Terminated
+    token: 8fmf94kf
+    links:
+      internal:
+        cert: <base64 encoded cert>
+        volumes:
+        - name: kubevirt-volume
+          - format: raw
+            url: https://virt-import.kubevirt/ns1/export-vm1/volumes/kubevirt-volume/raw?token=8fmf94kf
+          - format: archive
+            url: https://virt-import.kubevirt/ns1/export-vm1/volumes/kubevirt-volume/archive?token=8fmf94kf
+        - name: archive-volume
+          - format: dir
+            url: https://virt-import.kubevirt/ns1/export-vm1/volumes/archive-volume/dir?token=8fmf94kf
+          - format: archive
```

Comment:
I don't understand what are the different formats; Nor do I understand how would `status` look like if the VM has multiple disks, or once snapshots gain RAM.

Can you add an example of a multi-disk VM and how its VirtualMachineExport looks like? I'd like to ensure that I can easily recreate the VM on another cluster with access to its VirtualMachineExport alone.

Reply:
THe example has multiple disks.  One named `kubevirt-volume` and one named `archive-volume`.

Snapshots with ram has not been designed yet, but I think the idea is that there will be a special "memory" volume for VMs that is snapshotted/restored just like the others.  I expect "memory" volumes will be exported in the "raw" format described here.  Maybe "file" is a better name for that format?

Reply:
> THe example has multiple disks. One named kubevirt-volume and one named archive-volume.

I did not realize that. I would appreciate seeing a skeleton of the VM being exported in the proposal.

>  Maybe "file" is a better name for that format?

I'm not sure about the context of the question. "raw" tells me that the file is not compressed or otherwise treated. "file tells me less.

Reply:
yeah raw is better